packages/eslint-plugin/src/rules/no-misused-promises.ts

```diff
@@ -137,6 +158,164 @@ export default util.createRule<Options, 'conditional' | 'voidReturn'>({
       }
     }
 
+    function checkAssignments(node: TSESTree.AssignmentExpression): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      const varType = checker.getTypeAtLocation(tsNode.left);
+      if (!isVoidReturningFunctionType(checker, tsNode.left, varType)) {
+        return;
+      }
+
+      if (returnsThenable(checker, tsNode.right)) {
+        context.report({
+          messageId: 'voidReturnVariable',
+          node: node.right,
+        });
+      }
+    }
+
+    function checkVariableDeclaration(node: TSESTree.VariableDeclarator): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (tsNode.initializer === undefined || node.init === null) {
+        return;
+      }
+      const varType = checker.getTypeAtLocation(tsNode.name);
+      if (!isVoidReturningFunctionType(checker, tsNode.initializer, varType)) {
+        return;
+      }
+
+      if (returnsThenable(checker, tsNode.initializer)) {
+        context.report({
+          messageId: 'voidReturnVariable',
+          node: node.init,
+        });
+      }
+    }
+
+    function checkProperties(node: TSESTree.Property): void {
+      const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+      if (ts.isPropertyAssignment(tsNode)) {
+        const contextualType = checker.getContextualType(tsNode.initializer);
+        if (contextualType === undefined) {
+          return;
+        }
+        if (
+          !isVoidReturningFunctionType(
+            checker,
+            tsNode.initializer,
+            contextualType,
+          )
+        ) {
+          return;
+        }
+        if (returnsThenable(checker, tsNode.initializer)) {
+          context.report({
+            messageId: 'voidReturnProperty',
+            node: node.value,
+          });
+        }
+      } else if (ts.isShorthandPropertyAssignment(tsNode)) {
+        const contextualType = checker.getContextualType(tsNode.name);
+        if (contextualType === undefined) {
+          return;
+        }
+        if (
+          !isVoidReturningFunctionType(checker, tsNode.name, contextualType)
+        ) {
+          return;
+        }
+        if (returnsThenable(checker, tsNode.name)) {
+          context.report({
+            messageId: 'voidReturnProperty',
+            node: node.value,
+          });
+        }
+      } else if (ts.isMethodDeclaration(tsNode)) {
+        if (ts.isComputedPropertyName(tsNode.name)) {
+          return;
+        }
+        const obj = tsNode.parent;
+        if (!ts.isObjectLiteralExpression(obj)) {
+          return;
+        }
+        const objType =
+          checker.getContextualType(obj) ?? checker.getTypeAtLocation(obj);
```

Comment:
Out of curiosity: why would both of these be necessary? Most other places just use one or the other; what's the difference in behavior?

Reply:
I rechecked and `getTypeAtLocation` wasn't necessary here. I'm not sure what I was thinking then 😇 Thank you for pointing out!